File: DFS/291. Word Pattern II/291. Word Pattern II.java
class Solution {
    //dfs + backtracking
    // https://leetcode.com/problems/word-pattern-ii/discuss/73664/Share-my-Java-backtracking-solution
    public boolean wordPatternMatch(String pattern, String str) {
        Map<Character, String> map = new HashMap<>();
        Set<String> set = new HashSet<>();
        return isMatch(str, 0, pattern, 0, map, set);
    }
    private boolean isMatch(String str, int i, String pat, int j, Map<Character, String> map, Set<String> set) {
        // corner case
        if (i == str.length() && j == pat.length()) return true;
        if (i == str.length() || j == pat.length()) return false;
        
        char c = pat.charAt(j);
        if (map.containsKey(c)) {
            String cur = map.get(c);
            if (!str.startsWith(cur, i)) return false;
            
            return isMatch(str, i+cur.length(), pat, j+1, map, set);
        }
        // no exist
        for (int k = i; k < str.length(); k++) {
            String recur = str.substring(i, k+1);
            if (set.contains(recur)) {
                continue;
            }
            
            // create and update it
            map.put(c, recur);
            set.add(recur);
            if (isMatch(str, k+1, pat, j+1, map, set)) return true;
            // backtracking
            map.remove(c);
            set.remove(recur);
        }
        
        // we've tried out best but still not luck
        return false;
    }
}

File: DFS/291. Word Pattern II/291. Word Pattern II.cpp
class Solution {
public:
    // https://leetcode.com/problems/word-pattern-ii/discuss/73666/Share-my-C%2B%2B-backtracking-solution
    unordered_map<char, string>pDict;
    unordered_map<string, char>sDict;
    bool wordPatternMatch(string pattern, string str) {
        return isMatch(pattern, 0, str, 0);
    }
    bool isMatch(string &pattern, int i, string &str, int j) {
        // base case
        int m = pattern.size();
        int n = str.size();
        if (i == m || j == n) {
            if (i == m && j == n) return true;
            return false;
        }
        
        for (int k = j; k < n; k++) {
            bool ins = false;
            string cur = str.substr(j, k-j+1);
            
            // exist char related string
            if ( pDict.find(pattern[i]) != pDict.end() ) {
                if (pDict[pattern[i]] != cur) continue;
            } else if ( sDict.find(cur) != sDict.end() ) {
                if (sDict[cur] != pattern[i]) continue;
            } else {
                pDict[pattern[i]] = cur;
                sDict[cur] = pattern[i];
                ins = true;
            }
            if (isMatch(pattern, i+1, str, k+1)) return true;
            if (ins) { // backtracking
                pDict.erase(pattern[i]);
                sDict.erase(cur);
            }
        }
        
        return false;
    }
};
